sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,139 @@ def _sage_(self):
                                     hold=True)
         return f
 
+    def principal_value(self):
+        """
+        Compute the Cauchy Principal Value of the definite integral of a real function in the given interval
+        on the real axis.
+        In mathematics, the Cauchy principal value, is a method for assigning values to certain improper
+        integrals which would otherwise be undefined.
+        The Cauchy principal value of a finite integral of a function f about a point c with a <= c <= b is given by
+                b               c-r                       b
+              /                  /                      /
+             /                  /                      /
+             |                  |                      |
+             |                  |                      |
+        p.v. | f(x)dx  = lim    | f(x)dx   +   lim     |
+             |          r -> 0+ |             r -> 0+  |
+             |                  |                      |
+             /                  /                     /
+            /                  /                     /
+          a                 a                     c+r
+
+        Examples
+        ========
+
+        >>> from sympy import Dummy, symbols, integrate, limit, oo
+        >>> from sympy.integrals.integrals import Integral
+        >>> from sympy.calculus.singularities import singularities
+        >>> x = symbols('x')
+
+        >>> g = x + 1
+        >>> Integral(g, (x, -oo, oo)).principal_value()
+        oo
+
+        >>> f = 1/(x**3)
+        >>> singularities(f,x)
+        {0}
+
+        >>> Integral(f, (x, -oo, oo)).principal_value()
+        nan
+
+        This happens so because, the method principal_value returns -oo for Integral from -oo to 0
+        and +oo for Integral from 0 to oo. While adding both the Integrals i.e. -oo + oo we do not
+        get 0 (as we expect) but nan since sympy realises that -oo and +oo aren't actual numbers that
+        can be added/subtracted.
+
+        >>> Integral(f, (x, -oo, 0)).principal_value()
+        -oo
+
+        >>> Integral(f, (x, 0, oo)).principal_value()
+        oo
+
+        >>> Integral(f, (x, 1, 2)).principal_value()
+        3/8
+
+        >>> Integral(f, (x, -2, -1)).principal_value() + Integral(f, (x, 1, 2)).principal_value()
+        0
+
+        >>> Integral(f, (x, 0.4, 3)).principal_value()
+        3.06944444444444
+
+        >>> Integral(f, (x, -3, -0.4)).principal_value() + Integral(f, (x, 0.4, 3)).principal_value()
+        0
+
+
+        See also
+        ========
+
+        [1] https://en.wikipedia.org/wiki/Cauchy_principal_value
+        [2] http://mathworld.wolfram.com/CauchyPrincipalValue.html
+        """
+        from sympy.calculus import singularities
+
+        x, a, b = self.limits[0]
+
+        r = Dummy('r')
+
+        singularities_list = list(singularities(self.function, x))
```

Comment:
This could be the list of singularities in the interval (a, b): 

    [s for s in singularities(f, x) if s.is_comparable and a < s < b]

Reply:
@jksuom Yeah, that looks like a better idea. I'll implement it right away.